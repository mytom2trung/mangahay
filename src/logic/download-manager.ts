/* eslint-disable camelcase */
/* eslint-disable functional/no-throw-statement */
import hashSum from "hash-sum"
import type { Chapter, Comic, ComicChapter, RouteComic } from "raiku-pgs/plugin"
/*
.
├── meta/
│   ├── [hash_id_manga]/
│   │   └── [hash_id_ep].mod
│   └── [hash_id_manga].mod
├── files/
│   └── [hash_id_manga]
│       └── [hash_id_ep]
└── poster/
    └── [hash_id_manga]
*/

const DIR_META = "meta"
const DIR_POSTER = "poster"
const DIR_FILES = "files"

export interface ComicOnDisk extends Comic {
  readonly route: RouteComic
  readonly start_download_at: number
}
export interface ComicChapterOnDisk extends ComicChapter {
  readonly chapters: Chapter[]
  readonly route: RouteComic
  readonly ep_name: string
  readonly ep_param: string
  readonly downloaded: number
  readonly start_download_at: number
  readonly pages_offline: readonly string[]
}
export interface ComicChapterRunning extends ComicChapterOnDisk {
  downloaded: number
  readonly pages_offline: string[]
}

async function downloadFile(
  src: string,
  path: string,
  downloading: Ref<boolean>,
  retry: number
): Promise<void> {
  let buffer: ArrayBuffer

  const hashIndex = src.indexOf(HASH_TAG)
  if (hashIndex > -1) {
    let headers: Record<string, string> | undefined

    try {
      headers = JSON.parse(src.slice(hashIndex + HASH_TAG.length))
    } catch {}

    // request now
    buffer = await get({
      url: (src.startsWith("//") ? "https:" : "") + src.slice(0, hashIndex),
      headers,
      responseType: "arraybuffer"
    }).then((res) => base64ToUint8(res.data))
  } else {
    buffer = await await fetchRetry(fetch, [src], {
      retries: retry,
      retryDelay: 300
    }).then((res) => res.arrayBuffer())
  }

  if (!downloading.value) throw new Error("user_paused")
  await Filesystem.writeFile({
    path,
    data: uint8ToBase64(new Uint8Array(buffer)),
    directory: Directory.External,
    recursive: true
  })
}

async function downloadFiles(
  sources: readonly string[],
  hashIDManga: string,
  hashIDEp: string,
  downloading: Ref<boolean>,
  onprogress: (cur: number, total: number, path: string) => void
): Promise<void> {
  await someLimit(
    sources,
    async (src: string, index: number) => {
      if (src.startsWith(PROTOCOL_OFFLINE)) return false

      if (!downloading.value) {
        throw new Error("user_paused")
      }
      const path = `${DIR_FILES}/${hashIDManga}/${hashIDEp}/${hashSum(index)}`
      await downloadFile(src, path, downloading, 5)

      onprogress(index, sources.length, PROTOCOL_OFFLINE + path)

      return false
    },
    5
  )
}

async function saveMetaManga(
  route: RouteComic,
  metaManga: Comic
): Promise<ComicOnDisk> {
  const hash_id = hashSum(route.params.comic)

  const path = `${DIR_META}/${hash_id}.mod`

  // check
  try {
    const val = JSON.parse(
      await Filesystem.readFile({
        path,
        directory: Directory.External,
        encoding: Encoding.UTF8
      }).then((res) => res.data)
    )

    if (val) return val as ComicOnDisk
  } catch {}

  const pathPoster = `${DIR_POSTER}/${hash_id}`
  await downloadFile(
    metaManga.image,
    pathPoster,
    {
      value: true
    } as Ref<boolean>,
    5
  )

  const metaOnDisk: ComicOnDisk = {
    ...metaManga,
    route,
    image: `${PROTOCOL_OFFLINE}/${pathPoster}`,
    start_download_at: Date.now()
  }
  await Filesystem.writeFile({
    path,
    directory: Directory.External,
    encoding: Encoding.UTF8,
    data: JSON.stringify(metaOnDisk)
  })

  return metaOnDisk
}

export function createTaskDownloadEpisode(
  route: RouteComic,
  metaManga: Comic,
  metaEp: ComicChapter & {
    readonly chapters: Chapter[]
  },
  ep_name: string,
  ep_param: string,
  pages: readonly string[]
): {
  ref: ComicChapterRunning
  startSaveMetaManga: () => Promise<ComicOnDisk>
  downloading: globalThis.Ref<boolean>
  start: () => Promise<ComicChapterOnDisk | undefined>
  stop: () => Promise<ComicChapterOnDisk | undefined>
  resume: () => Promise<ComicChapterOnDisk | undefined>
} {
  const hashIDManga = hashSum(route.params.comic)
  const hashIDEp = hashSum(ep_param)

  const downloading = ref(false)
  const refValue = reactive<ComicChapterRunning>({
    start_download_at: Date.now(),
    downloaded: 0,
    ...metaEp,
    route,
    pages_offline: pages.slice(0),
    ep_name,
    ep_param
  })

  const startSaveMetaManga = () => saveMetaManga(route, metaManga)

  let timeout: NodeJS.Timeout | number
  let taskSaveMeta: Promise<ComicChapterOnDisk> | undefined
  const saveMeta = (metaCloned: ComicChapterOnDisk) => {
    taskSaveMeta = new Promise<ComicChapterOnDisk>((resolve, reject) => {
      // delay 1s
      clearTimeout(timeout)

      timeout = setTimeout(async () => {
        try {
          await Filesystem.writeFile({
            path: `${DIR_META}/${hashIDManga}/${hashIDEp}.mod`,
            data: JSON.stringify(metaCloned),
            directory: Directory.External,
            encoding: Encoding.UTF8
          })
          resolve(metaCloned)
        } catch (err) {
          reject(err)
        }
      }, 1_000)
    })
    return taskSaveMeta
  }

  const start = async () => {
    if (downloading.value) {
      console.warn("task is running")
      return
    }

    downloading.value = true
    const hashIDManga = hashSum((await startSaveMetaManga()).route.params.comic)

    // check continue this passed
    const metaInDisk = await Filesystem.readFile({
      path: `${DIR_META}/${hashIDManga}/${hashIDEp}.mod`,
      directory: Directory.External,
      encoding: Encoding.UTF8
    })
      .then(
        (res) =>
          JSON.parse(res.data) as ComicChapterOnDisk & {
            downloaded: number
          }
      )
      .catch(() => undefined)

    // save meta
    Object.assign(refValue, metaInDisk)

    if (!downloading.value) return

    // save files
    await downloadFiles(
      refValue.pages_offline,
      hashIDManga,
      hashIDEp,
      downloading,
      (cur, total, path) => {
        refValue.pages_offline[cur] = path
        refValue.downloaded++
        void saveMeta(refValue)
      }
    ).catch(async (err) => {
      await saveMeta(refValue)
      downloading.value = false
      throw err
    })

    const meta = await saveMeta(refValue)
    downloading.value = false
    return meta
  }
  const stop = async () => {
    downloading.value = false
    return taskSaveMeta
  }
  const resume = start

  return { ref: refValue, startSaveMetaManga, downloading, start, stop, resume }
}

export async function getListManga() {
  try {
    // return
    const { files } = await Filesystem.readdir({
      path: DIR_META,
      directory: Directory.External
    })

    const list = (
      await Promise.all(
        // eslint-disable-next-line array-callback-return
        files.map((item) => {
          if (item.type === "file")
            return Filesystem.readFile({
              path: `${DIR_META}/${item.name}`,
              directory: Directory.External,
              encoding: Encoding.UTF8
            }).then((res) => JSON.parse(res.data) as ComicOnDisk)
          // .catch(() => null)
        })
      )
    ).filter(Boolean) as ComicOnDisk[]

    list.sort((a, b) => a.start_download_at - b.start_download_at)

    return list
  } catch (err) {
    if ((err as Error | undefined)?.message === "Folder does not exist.")
      return []

    throw err
  }
}

export async function getCountEpisodes(comic: string) {
  const hashIDManga = hashSum(comic)

  try {
    const { files } = await Filesystem.readdir({
      path: `${DIR_META}/${hashIDManga}`,
      directory: Directory.External
    })

    return files.length
  } catch (err) {
    if ((err as Error | undefined)?.message === "Folder does not exist.")
      return 0

    throw err
  }
}

export async function getListEpisodes(comic: string) {
  const hashIDManga = hashSum(comic)

  const { files } = await Filesystem.readdir({
    path: `${DIR_META}/${hashIDManga}`,
    directory: Directory.External
  })

  return Promise.all(
    files.map(
      (item) =>
        item.type === "file" &&
        Filesystem.readFile({
          path: `${DIR_META}/${hashIDManga}/${item.name}`,
          directory: Directory.External,
          encoding: Encoding.UTF8
        }).then((res) => JSON.parse(res.data) as ComicChapterOnDisk)
    )
  ).then((list) => list.filter(Boolean) as ComicChapterOnDisk[])
}

export async function deleteManga(comic: string) {
  const hashIDManga = hashSum(comic)

  await Promise.all([
    // remove meta episodes
    Filesystem.rmdir({
      path: `${DIR_META}/${hashIDManga}`,
      directory: Directory.External,
      recursive: true
    }).catch(() => null),

    // remove meta manga
    Filesystem.deleteFile({
      path: `${DIR_META}/${hashIDManga}.mod`,
      directory: Directory.External
    }).catch(() => null),

    // remove poster
    Filesystem.deleteFile({
      path: `${DIR_POSTER}/${hashIDManga}`,
      directory: Directory.External
    }).catch(() => null),

    // remove pages of manga
    Filesystem.rmdir({
      path: `${DIR_FILES}/${hashIDManga}`,
      directory: Directory.External,
      recursive: true
    }).catch(() => null)
  ])
}

export async function deleteEpisode(comic: string, ep_param: string) {
  const hashIDManga = hashSum(comic)
  const hashIDEp = hashSum(ep_param)

  await Promise.all([
    // remove meta
    Filesystem.deleteFile({
      path: `${DIR_META}/${hashIDManga}/${hashIDEp}.mod`,
      directory: Directory.External
    })
      .then(async () => {
        // check removed all episode
        const { files } = await Filesystem.readdir({
          path: `${DIR_META}/${hashIDManga}`,
          directory: Directory.External
        })

        if (files.length === 0) {
          // remove poster and meta manga
          await Promise.all([
            // remove poster
            Filesystem.deleteFile({
              path: `${DIR_POSTER}/${hashIDManga}`,
              directory: Directory.External
              // eslint-disable-next-line promise/no-nesting
            }).catch(() => null),
            // remove meta manga
            Filesystem.deleteFile({
              path: `${DIR_META}/${hashIDManga}.mod`,
              directory: Directory.External
              // eslint-disable-next-line promise/no-nesting
            }).catch(() => null)
          ])
        }

        // eslint-disable-next-line no-useless-return
        return
      })
      .catch(() => null),

    // remove pages
    Filesystem.rmdir({
      path: `${DIR_FILES}/${hashIDManga}/${hashIDEp}`,
      directory: Directory.External,
      recursive: true
    })
      .then(async () => {
        // check removed all episode
        const { files } = await Filesystem.readdir({
          path: `${DIR_FILES}/${hashIDManga}`,
          directory: Directory.External
        })

        if (files.length === 0) {
          await Filesystem.rmdir({
            path: `${DIR_FILES}/${hashIDManga}`,
            directory: Directory.External,
            recursive: true
          })
        }

        // eslint-disable-next-line no-useless-return
        return
      })
      .catch(() => null)
  ])
}

export async function getComic(comic: string) {
  const hashIDManga = hashSum(comic)

  return JSON.parse(
    await Filesystem.readFile({
      path: `${DIR_META}/${hashIDManga}.mod`,
      directory: Directory.External,
      encoding: Encoding.UTF8
    }).then((res) => res.data)
  ) as ComicOnDisk
}

export async function getEpisode(comic: string, ep_param: string) {
  const hashIDManga = hashSum(comic)
  const hashIDEp = hashSum(ep_param)

  return JSON.parse(
    await Filesystem.readFile({
      path: `${DIR_META}/${hashIDManga}/${hashIDEp}.mod`,
      directory: Directory.External,
      encoding: Encoding.UTF8
    }).then((res) => res.data)
  ) as ComicChapterOnDisk
}

// eslint-disable-next-line functional/no-mixed-type
export interface TaskDDEp {
  ref: ComicChapterOnDisk
  downloading?: boolean
  stop?: () => void
}
export type TaskDLEp = Pick<
  ReturnType<typeof createTaskDownloadEpisode>,
  "ref" | "downloading" | "stop" | "resume"
>

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function isTaskDLEp(value: any): value is TaskDLEp {
  return typeof value?.resume === "function"
}
