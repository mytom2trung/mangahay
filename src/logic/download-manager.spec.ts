/* eslint-disable camelcase */
import hashSum from "hash-sum"
import type { Chapter, Comic, ComicChapter, RouteComic } from "raiku-pgs/plugin"
import { cleanup, exists, readdir, readFile } from "test/vitest/utils"
import { expect, vi } from "vitest"
import createFetchMock from "vitest-fetch-mock"

import { createTaskDownloadEpisode, getListManga } from "./download-manager"

const fetchMocker = createFetchMock(vi)

Object.assign(global, { fetch: fetchMocker })
global.Date.now = vi.fn()

const manga_id = "1"
const manga_name = "Hôn Nhân Hạnh Phúc Của Tôi"
const manga_image = "http://localhost/poster/manga-1.jpg"
const source_id = "nettruyen"
const ep_id = "1234"
const ep_name = "Chapter 1"
const ep_param = "chap-1-i1234"
const pages = [
  "https://localhost/pages/1.png",
  "https://localhost/pages/2.png",
  "https://localhost/pages/3.png",
  "https://localhost/pages/4.png",
  "https://localhost/pages/5.png",
  "https://localhost/pages/6.png",
  "https://localhost/pages/7.png",
  "https://localhost/pages/8.png"
]

const route: RouteComic = {
  name: "comic",
  params: {
    sourceId: source_id,
    comic: "manga-1"
  }
}
const metaManga: Comic = {
  name: manga_name,
  othername: "",
  manga_id,
  updated_at: 1682580180000,
  likes: null,
  image: manga_image,
  author: [
    {
      name: "Agoki Akumi",
      route: {
        name: "author",
        params: { sourceId: "nettruyen", type: "" },
        query: { "tac-gia": "Agoki Akumi" }
      }
    }
  ],
  status: "Đang tiến hành",
  genres: [],
  views: 1626445,
  rate: { cur: 4.4, max: 5, count: 2032 },
  follows: 14821,
  description: "",
  chapters: [
    {
      id: "776047",
      route: {
        name: "comic chap",
        params: {
          sourceId: "nettruyen",
          comic: "hon-nhan-hanh-phuc-cua-toi",
          chap: "20-2-i776047"
        }
      },
      name: "20",
      updated_at: 1634169600000,
      views: 62085
    }
  ],
  comments: [],
  comments_count: 1,
  comments_pages: 1
}
const metaEp: ComicChapter & {
  chapters: Chapter[]
} = {
  name: manga_id,
  image: manga_image,
  manga_id,
  path_manga: {
    name: "comic",
    params: {
      comic: "hon-nhan-hanh-phuc-cua-toi-25737",
      sourceId: "nettruyen"
    }
  },
  ep_id,
  updated_at: 1634183940000,
  pages: pages.map((src) => ({ src })),
  cdn: "//cdn.ntcdntempv26.com",
  cdn2: "//cdn.ntcdntempv3.com",
  chapters: <Chapter[]>[],
  comments: [],
  comments_count: 29,
  comments_pages: 1
}

const hashIDManga = hashSum(route.params.comic)
const hashIDEp = hashSum(ep_param)

function patchFetch() {
  // continue download
  ;(fetch as ReturnType<typeof vi.fn>).mockReset()
  ;(fetch as ReturnType<typeof vi.fn>).mockImplementation(
    async (url: string) => {
      await sleep(100)
      return Promise.resolve({
        async arrayBuffer() {
          return new TextEncoder().encode(url)
        },

        async text() {
          return url
        }
      })
    }
  )
}
;(Date.now as ReturnType<typeof vi.fn>).mockReturnValue(0)
fetchMocker.enableMocks()

describe("download-manager", () => {
  beforeEach(async () => {
    await cleanup()
    await sleep(1000)
  })

  test("should download episode x for the first time", async () => {
    fetchMocker.mockResponse(({ url }) => url)

    const { ref, start, downloading } = createTaskDownloadEpisode(
      route,
      metaManga,
      metaEp,
      ep_name,
      ep_param,
      pages
    )
    expect(ref.downloaded).toBe(0)
    expect(downloading.value).toBeFalsy()

    const watcher = vi.fn()
    watch(ref, watcher, { deep: true })

    await start()

    // check directory
    await expect(readdir("")).resolves.toEqual(["files", "meta", "poster"])

    await expect(readdir("files")).resolves.toEqual([hashIDManga])
    await expect(readdir("files/" + hashIDManga)).resolves.toEqual([hashIDEp])

    await expect(readdir("meta")).resolves.toEqual([
      hashIDManga,
      hashIDManga + ".mod"
    ])
    await expect(readdir("meta/" + hashIDManga)).resolves.toEqual([
      hashIDEp + ".mod"
    ])

    await expect(readdir("poster")).resolves.toEqual([hashIDManga])

    expect(await readdir(`files/${hashIDManga}/${hashIDEp}`)).toEqual([
      "1a96284a",
      "1a96284b",
      "1a96284c",
      "1a96284d",
      "1a96284e",
      "1a96284f",
      "1a962850",
      "1a962851"
    ])

    // valid image pages

    for (const index in pages) {
      const path = `files/${hashIDManga}/${hashIDEp}/${hashSum(+index)}`

      expect(await readFile(path, Encoding.UTF8)).toBe(pages[index])
    }

    // valid meta manga
    expect(
      JSON.parse(await readFile(`meta/${hashIDManga}.mod`, Encoding.UTF8))
    ).toEqual({
      ...metaManga,
      route,
      image: "offline:///poster/4523dfa2",
      start_download_at: 0
    })

    // valid meta episode
    expect(
      JSON.parse(
        await readFile(`meta/${hashIDManga}/${hashIDEp}.mod`, Encoding.UTF8)
      )
    ).toEqual({
      ...metaEp,
      ep_name,
      ep_param,
      route,
      start_download_at: 0,
      downloaded: 8,
      pages_offline: [
        "offline://files/4523dfa2/62c5450c/1a96284a",
        "offline://files/4523dfa2/62c5450c/1a96284b",
        "offline://files/4523dfa2/62c5450c/1a96284c",
        "offline://files/4523dfa2/62c5450c/1a96284d",
        "offline://files/4523dfa2/62c5450c/1a96284e",
        "offline://files/4523dfa2/62c5450c/1a96284f",
        "offline://files/4523dfa2/62c5450c/1a962850",
        "offline://files/4523dfa2/62c5450c/1a962851"
      ]
    })

    // valid meta image
    expect(await readFile("poster/" + hashIDManga, Encoding.UTF8)).toBe(
      manga_image
    )
    expect(watcher.mock.calls.length).toBe(8)
  })

  test("should forcibly stopped while downloading", async () => {
    let counter = 0
    fetchMocker.mockResponse(async ({ url }) => {
      // await sleep(500)
      if (counter++ > 5) return Promise.reject(new Error("time_out"))

      return url
    })

    const { ref, start, downloading } = createTaskDownloadEpisode(
      route,
      metaManga,
      metaEp,
      ep_name,
      ep_param,
      pages
    )
    expect(ref.downloaded).toBe(0)
    expect(downloading.value).toBeFalsy()

    await start().catch(() => null)

    // check hash file page
    expect(await readdir(`files/${hashIDManga}/${hashIDEp}`)).toEqual([
      "1a96284a",
      "1a96284b",
      "1a96284c",
      "1a96284d",
      "1a96284e"
    ])

    // valid image page

    for (const index in pages.slice(0, 5)) {
      const path = `files/${hashIDManga}/${hashIDEp}/${hashSum(+index)}`

      expect(await readFile(path, Encoding.UTF8)).toBe(pages[index])
    }

    // valid meta
    expect(
      JSON.parse(
        await readFile(
          "meta/" + hashIDManga + "/" + hashIDEp + ".mod",
          Encoding.UTF8
        )
      )
    ).toEqual({
      ...metaEp,
      ep_name,
      ep_param,
      route,
      start_download_at: 0,
      downloaded: 5,
      pages_offline: [
        "offline://files/4523dfa2/62c5450c/1a96284a",
        "offline://files/4523dfa2/62c5450c/1a96284b",
        "offline://files/4523dfa2/62c5450c/1a96284c",
        "offline://files/4523dfa2/62c5450c/1a96284d",
        "offline://files/4523dfa2/62c5450c/1a96284e",
        "https://localhost/pages/6.png",
        "https://localhost/pages/7.png",
        "https://localhost/pages/8.png"
      ]
    })
  })

  test("should continue while downloading", async () => {
    let counter = 0
    fetchMocker.mockResponse(async ({ url }) => {
      // await sleep(500)
      if (counter++ > 5) return Promise.reject(new Error("time_out"))

      return url
    })

    const { ref, start, downloading } = createTaskDownloadEpisode(
      route,
      metaManga,
      metaEp,
      ep_name,
      ep_param,
      pages
    )
    expect(ref.downloaded).toBe(0)
    expect(downloading.value).toBeFalsy()

    await start().catch(() => null)
    patchFetch()

    // check hash file page
    expect(await readdir(`files/${hashIDManga}/${hashIDEp}`)).toEqual([
      "1a96284a",
      "1a96284b",
      "1a96284c",
      "1a96284d",
      "1a96284e"
    ])

    // valid image page

    for (const index in pages.slice(0, 5)) {
      const path = `files/${hashIDManga}/${hashIDEp}/${hashSum(+index)}`

      expect(await readFile(path, Encoding.UTF8)).toBe(pages[index])
    }

    // valid meta
    expect(
      JSON.parse(
        await readFile(
          "meta/" + hashIDManga + "/" + hashIDEp + ".mod",
          Encoding.UTF8
        )
      )
    ).toEqual({
      ...metaEp,
      ep_name,
      ep_param,
      route,
      start_download_at: 0,
      downloaded: 5,
      pages_offline: [
        "offline://files/4523dfa2/62c5450c/1a96284a",
        "offline://files/4523dfa2/62c5450c/1a96284b",
        "offline://files/4523dfa2/62c5450c/1a96284c",
        "offline://files/4523dfa2/62c5450c/1a96284d",
        "offline://files/4523dfa2/62c5450c/1a96284e",
        "https://localhost/pages/6.png",
        "https://localhost/pages/7.png",
        "https://localhost/pages/8.png"
      ]
    })

    const {
      ref: ref2,
      start: start2,
      downloading: dl2
    } = createTaskDownloadEpisode(
      route,
      metaManga,
      metaEp,
      ep_name,
      ep_param,
      pages
    )
    expect(ref2.downloaded).toBe(0)
    expect(dl2.value).toBeFalsy()

    await start2().catch(() => null)

    // check hash file page
    expect(await readdir(`files/${hashIDManga}/${hashIDEp}`)).toEqual([
      "1a96284a",
      "1a96284b",
      "1a96284c",
      "1a96284d",
      "1a96284e",
      "1a96284f",
      "1a962850",
      "1a962851"
    ])

    // valid image pages

    for (const index in pages) {
      const path = `files/${hashIDManga}/${hashIDEp}/${hashSum(+index)}`

      expect(await readFile(path, Encoding.UTF8)).toBe(pages[index])
    }

    // valid meta
    expect(
      JSON.parse(
        await readFile(
          "meta/" + hashIDManga + "/" + hashIDEp + ".mod",
          Encoding.UTF8
        )
      )
    ).toEqual({
      ...metaEp,
      ep_name,
      ep_param,
      route,
      start_download_at: 0,
      downloaded: 8,
      pages_offline: [
        "offline://files/4523dfa2/62c5450c/1a96284a",
        "offline://files/4523dfa2/62c5450c/1a96284b",
        "offline://files/4523dfa2/62c5450c/1a96284c",
        "offline://files/4523dfa2/62c5450c/1a96284d",
        "offline://files/4523dfa2/62c5450c/1a96284e",
        "offline://files/4523dfa2/62c5450c/1a96284f",
        "offline://files/4523dfa2/62c5450c/1a962850",
        "offline://files/4523dfa2/62c5450c/1a962851"
      ]
    })
  })

  // test("should run multiple download episodes", async () => {})

  test("should download stop and resume", async () => {
    patchFetch()

    const { ref, downloading, start, stop, resume } = createTaskDownloadEpisode(
      route,
      metaManga,
      metaEp,
      ep_name,
      ep_param,
      pages
    )
    expect(downloading.value).toBe(false)
    expect(ref.downloaded).toBe(0)

    const watcher = vi.fn()
    watch(ref, watcher, { deep: true })

    void start()
    await sleep(500)
    expect(downloading.value).toBe(true)
    expect(ref.downloaded).toBeGreaterThanOrEqual(1)

    void stop()
    await sleep(500)

    expect(downloading.value).toBe(false)
    expect(ref.downloaded).toBeGreaterThanOrEqual(1)

    const promise = resume()
    expect(downloading.value).toBe(true)
    expect(ref.downloaded).toBeGreaterThanOrEqual(1)

    await promise

    // check hash file page
    expect(await readdir(`files/${hashIDManga}/${hashIDEp}`)).toEqual([
      "1a96284a",
      "1a96284b",
      "1a96284c",
      "1a96284d",
      "1a96284e",
      "1a96284f",
      "1a962850",
      "1a962851"
    ])

    // valid image pages

    for (const index in pages) {
      const path = `files/${hashIDManga}/${hashIDEp}/${hashSum(+index)}`

      expect(await readFile(path, Encoding.UTF8)).toBe(pages[index])
    }

    // valid meta
    expect(
      JSON.parse(
        await readFile(
          "meta/" + hashIDManga + "/" + hashIDEp + ".mod",
          Encoding.UTF8
        )
      )
    ).toEqual({
      ...metaEp,
      ep_name,
      ep_param,
      route,
      start_download_at: 0,
      downloaded: 8,
      pages_offline: [
        "offline://files/4523dfa2/62c5450c/1a96284a",
        "offline://files/4523dfa2/62c5450c/1a96284b",
        "offline://files/4523dfa2/62c5450c/1a96284c",
        "offline://files/4523dfa2/62c5450c/1a96284d",
        "offline://files/4523dfa2/62c5450c/1a96284e",
        "offline://files/4523dfa2/62c5450c/1a96284f",
        "offline://files/4523dfa2/62c5450c/1a962850",
        "offline://files/4523dfa2/62c5450c/1a962851"
      ]
    })
  })

  test("should get list manga downloaded", async () => {
    await createTaskDownloadEpisode(
      route,
      metaManga,
      metaEp,
      ep_name,
      ep_param,
      pages
    ).start()

    // ok get list

    expect(await getListManga()).toEqual([
      {
        ...metaManga,
        route,
        image: "offline:///poster/4523dfa2",
        start_download_at: 0
      }
    ])

    const meta2 = {
      ...metaManga
    }

    await createTaskDownloadEpisode(
      {
        ...route,
        params: {
          ...route.params,
          comic: "runn"
        }
      },
      meta2,
      metaEp,
      ep_name,
      ep_param,
      pages
    ).start()

    expect(await getListManga()).toEqual([
      {
        ...metaManga,
        route: {
          ...route,
          params: {
            ...route.params,
            comic: "runn"
          }
        },
        manga_id: "1",
        image: "offline:///poster/2d293839",
        start_download_at: 0
      },
      {
        ...metaManga,
        route,
        image: "offline:///poster/4523dfa2",
        start_download_at: 0
      }
    ])
  })

  test("should get list episodes downloaded", async () => {
    await createTaskDownloadEpisode(
      route,
      metaManga,
      metaEp,
      ep_name,
      ep_param,
      pages
    ).start()

    // ok get list
    expect(await getListEpisodes(route.params.comic)).toEqual([
      {
        ...metaEp,
        ep_name,
        ep_param,
        route,
        start_download_at: 0,
        downloaded: 8,
        pages_offline: [
          "offline://files/4523dfa2/62c5450c/1a96284a",
          "offline://files/4523dfa2/62c5450c/1a96284b",
          "offline://files/4523dfa2/62c5450c/1a96284c",
          "offline://files/4523dfa2/62c5450c/1a96284d",
          "offline://files/4523dfa2/62c5450c/1a96284e",
          "offline://files/4523dfa2/62c5450c/1a96284f",
          "offline://files/4523dfa2/62c5450c/1a962850",
          "offline://files/4523dfa2/62c5450c/1a962851"
        ]
      }
    ])
    ;(Date.now as ReturnType<typeof vi.fn>).mockReturnValue(1)

    await createTaskDownloadEpisode(
      route,
      metaManga,
      { ...metaEp },
      ep_name,
      ep_param + 1,
      pages
    ).start()

    expect(await getListEpisodes(route.params.comic)).toEqual([
      {
        ...metaEp,
        route,
        ep_name,
        ep_param: ep_param + 1,
        downloaded: 8,
        ep_id,
        pages_offline: [
          `offline://files/${hashIDManga}/143946b6/1a96284a`,
          `offline://files/${hashIDManga}/143946b6/1a96284b`,
          `offline://files/${hashIDManga}/143946b6/1a96284c`,
          `offline://files/${hashIDManga}/143946b6/1a96284d`,
          `offline://files/${hashIDManga}/143946b6/1a96284e`,
          `offline://files/${hashIDManga}/143946b6/1a96284f`,
          `offline://files/${hashIDManga}/143946b6/1a962850`,
          `offline://files/${hashIDManga}/143946b6/1a962851`
        ],
        start_download_at: 1
      },
      {
        ...metaEp,
        route,
        ep_name,
        ep_param,
        start_download_at: 0,
        downloaded: 8,
        pages_offline: [
          `offline://files/${hashIDManga}/${hashIDEp}/1a96284a`,
          `offline://files/${hashIDManga}/${hashIDEp}/1a96284b`,
          `offline://files/${hashIDManga}/${hashIDEp}/1a96284c`,
          `offline://files/${hashIDManga}/${hashIDEp}/1a96284d`,
          `offline://files/${hashIDManga}/${hashIDEp}/1a96284e`,
          `offline://files/${hashIDManga}/${hashIDEp}/1a96284f`,
          `offline://files/${hashIDManga}/${hashIDEp}/1a962850`,
          `offline://files/${hashIDManga}/${hashIDEp}/1a962851`
        ]
      }
    ])
    ;(Date.now as ReturnType<typeof vi.fn>).mockReturnValue(0)
  })

  test("should delete manga", async () => {
    await createTaskDownloadEpisode(
      route,
      metaManga,
      metaEp,
      ep_name,
      ep_param,
      pages
    ).start()

    await expect(exists(`meta/${hashIDManga}`)).resolves.toBeTruthy()
    await expect(exists(`poster/${hashIDManga}`)).resolves.toBeTruthy()
    await expect(exists(`files/${hashIDManga}`)).resolves.toBeTruthy()

    await deleteManga(route.params.comic)

    await expect(exists(`meta/${hashIDManga}`)).resolves.toBeFalsy()
    await expect(exists(`poster/${hashIDManga}`)).resolves.toBeFalsy()
    await expect(exists(`files/${hashIDManga}`)).resolves.toBeFalsy()
  })

  test("should delete episode", async () => {
    await createTaskDownloadEpisode(
      route,
      metaManga,
      metaEp,
      ep_name,
      ep_param,
      pages
    ).start()
    await createTaskDownloadEpisode(
      route,
      metaManga,
      { ...metaEp, ep_id: metaEp.ep_id + 1 + "" },
      ep_name,
      ep_param + 1,
      pages
    ).start()

    await expect(exists(`meta/${hashIDManga}`)).resolves.toBeTruthy()
    await expect(exists(`poster/${hashIDManga}`)).resolves.toBeTruthy()
    await expect(exists(`files/${hashIDManga}`)).resolves.toBeTruthy()

    await expect(
      exists(`meta/${hashIDManga}/${hashIDEp}.mod`)
    ).resolves.toBeTruthy()
    await expect(
      exists(`files/${hashIDManga}/${hashIDEp}`)
    ).resolves.toBeTruthy()

    const hashIDEp2 = hashSum(ep_param + 1)
    await expect(
      exists(`meta/${hashIDManga}/${hashIDEp2}.mod`)
    ).resolves.toBeTruthy()
    await expect(
      exists(`files/${hashIDManga}/${hashIDEp2}`)
    ).resolves.toBeTruthy()

    await deleteEpisode(route.params.comic, ep_param)

    await expect(
      exists(`meta/${hashIDManga}/${hashIDEp}.mod`)
    ).resolves.toBeFalsy()
    await expect(
      exists(`files/${hashIDManga}/${hashIDEp}`)
    ).resolves.toBeFalsy()
    await expect(
      exists(`meta/${hashIDManga}/${hashIDEp2}.mod`)
    ).resolves.toBeTruthy()
    await expect(
      exists(`files/${hashIDManga}/${hashIDEp2}`)
    ).resolves.toBeTruthy()

    await deleteEpisode(route.params.comic, ep_param + 1)

    await expect(exists(`meta/${hashIDManga}.mod`)).resolves.toBeFalsy()
    await expect(exists(`poster/${hashIDManga}`)).resolves.toBeFalsy()
    await expect(exists(`files/${hashIDManga}`)).resolves.toBeFalsy()
  })
})

export {}
