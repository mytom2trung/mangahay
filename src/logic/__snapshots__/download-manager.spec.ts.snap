// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`download-manager > should download episode x for the first time 1`] = `
{
  "downloaded": 8,
  "ep_id": 1234,
  "ep_name": "Chapter 1",
  "pages": [
    "offline://files/1a96284b/30089c30/1a96284a",
    "offline://files/1a96284b/30089c30/1a96284b",
    "offline://files/1a96284b/30089c30/1a96284c",
    "offline://files/1a96284b/30089c30/1a96284d",
    "offline://files/1a96284b/30089c30/1a96284e",
    "offline://files/1a96284b/30089c30/1a96284f",
    "offline://files/1a96284b/30089c30/1a962850",
    "offline://files/1a96284b/30089c30/1a962851",
  ],
  "start_download_at": 1690022500169,
}
`;

exports[`download-manager > should forcibly stopped while downloading 1`] = `"time_out"`;
